text/0045-apply-keyword.md

```diff
@@ -0,0 +1,119 @@
+# Add Apply keyword to Cedar language
+
+## Related issues and PRs
+
+- Reference Issues: 
+- Implementation PR(s): 
+
+## Timeline
+
+- Started: 2024-01-25
+
+## Summary
+
+Add a keyword that allows denoted policy statements to have no impact on the
+authorization decision, yet be logged in the diagnostics for later annotation
+aggregation.
+
+## Basic example
+
+```cedar
+@annotation("value")
+apply (
+    principal,
+    action,
+    resource
+);
+```
+
+## Motivation
+
+I want to be able to create policy statements that will add annotations while
+not affecting the authorization outcome.  Here are some examples:
+
+1. I want to be able to require MFA anytime someone accesses a certain table.
+However, I want to be able to define permission for this table elsewhere.
+
+2. I want to be able to send out an email everytime something is denied.
+However, I want to be able to define what things are forbidden elsewhere.
+
+3. I want to be able to send out an email regardless of if a specific event
+is allowed or denied.
+
+## Detailed design
+
+There are three cases that are added to the language syntax:
+
+1. Apply on permit
+
+```cedar
+@mfa("true")
+apply on permit (
+    principal,
+    action == Postgres::Action::"select",
+    resource == Postgres::Table::"example.com:5432/db"
+);
+```
+
+2. Apply on forbid
+
+```cedar
+@email("alert@example.com")
+apply on forbid (
+    principal,
+    action,
+    resource
+);
+```
+
+3. Always apply
+
+```cedar
+@email("alert@example.com")
+apply (
+    principal,
+    action,
+    resource in Postgres::Table::"example.com:5432/db"
+);
+```
+
+The apply keyword will cause all matching policy statements to be added to the
+diagnostics, while not impacting the authorization decision in any way.
+
+### Changes required to parser/AST
+
+- Update grammer to support `apply on permit` `apply on forbid` `apply` in
+  addition to the existing `forbid` and `permit`
+
+### Changes required to evaluator
+
+- Policies that are categorized as `apply` would not impact the decision, while matches
+would be added to the diagnostic.
+
+## Drawbacks
+
+Why should we *not* do this? Please consider:
+
+- this does add the `apply` and `on` keyword to the grammar, thus increasing the
+complexity of the Cedar language
+- this may have some impact on the ability to short-circuit policy decisions,
+since all apply statements need to be checked to be added to the diagnostics
+
+## Alternatives
```

Comment:
Can you also add some suggestions for alternative names? I'm personally partial to `log`, `info`, or `advise`, since we use "apply" in other contexts (e.g., to say that a policy "applies" for a particular request).

Reply:
Thanks, I've added a list of alternative name possibilities to the Alternatives section.